scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -193,6 +194,108 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
+      s(side).getOrElse(kv._1, Iterable()).toSeq.map(w => (kv._1, (kv._2, w)))
+    }.toSCollection
+  }
+
+  /**
+   * Perform a left outer join by replicating `that` to all workers. The right side should be tiny
+   * and fit in memory.
+   *
+   * @group join
+   */
+  def hashLeftJoin[W: ClassTag](that: SCollection[(K, W)])
+  : SCollection[(K, (V, Option[W]))] = self.transform { in =>
+    val side = that.asMultiMapSideInput
+    in.withSideInputs(side).flatMap[(K, (V, Option[W]))] { (kv, s) =>
+      val (k, v) = kv
+      val m = s(side)
+      if (m.contains(k)) m(k).map(w => (k, (v, Some(w)))) else Seq((k, (v, None)))
+    }.toSCollection
+  }
+
+  /* Skew operations */
+
+  /**
+   * N to 1 skewproof flavor of [[PairSCollectionFunctions.join()]].
+   *
+   * This flavor of join, will try to reduce the skewness by performing two step join, on hot keys,
+   * and the rest. Number of values for a given key is estimated, with `1 - delta` probability,
+   * and the estimate is within `eps * N` of the true frequency
+   * (i.e., `true frequency <= estimate <= true frequency + eps * N`), where N is the total size of
+   * the left hand side stream so far.
+   *
+   * @note Make sure to import [[com.twitter.algebird.CMSHasherImplicits]] before using this join
+   *
+   * @example {{{
+   * // Implicits that enabling CMS-hashing
+   * import com.twitter.algebird.CMSHasherImplicits._
+   *
+   * val p = logs.skewedN1Join(logMetadata, hotKeyThreshold = 9000)
+   *
+   * }}}
+   *
+   * Read more about CMS -> [[com.twitter.algebird.CMSMonoid]]
+   *
+   * @group join
+   *
+   * @param hotKeyThreshold key with `hotKeyThreshold` values will be considered hot
+   * @param eps One-sided error bound on the error of each point query, i.e. frequency estimate.
+   * @param delta A bound on the probability that a query estimate does not lie within some small
+   *              interval (an interval that depends on `eps`) around the truth.
+   * @param seed A seed to initialize the random number generator used to create the pairwise
+   *             independent hash functions.
+   */
+  def skewedN1Join[W: ClassTag](that: SCollection[(K, W)],
+                                hotKeyThreshold: Long = 9000,
+                                eps: Double = 0.0001,
+                                delta: Double = 1E-10,
+                                seed: Int = 42)(implicit hasher: CMSHasher[K])
+  : SCollection[(K, (V, W))] = {
+    import com.twitter.algebird._
+
+    // Count k->#values:
+    val keyAggregator = CMS.aggregator[K](eps, delta, seed)
+
+    // Use asIterableSideInput as workaround for:
+    // http://stackoverflow.com/questions/37126729/ismsinkwriter-expects-keys-to-be-written-in-strictly-increasing-order
+    val keyCMS = self.keys.aggregate(keyAggregator).asIterableSideInput
+
+    val (hotSelf, chillSelf) = (SideOutput[(K,V)](), SideOutput[(K,V)]())
+    val partitionedSelf = self
```

Comment:
Isn't this gonna blow up left site? If I understand it correctly it will duplicate whole CMS for each record of the left side. While partition is basically a filter on CMS?